integration_tests/test_authz.py

```diff
@@ -0,0 +1,382 @@
+        cluster.balance(receiver_address)
+        == receiver_initial_balance + transaction_coins
+    )
+
+
+def test_execute_delegate_to_allowed_validator(cluster, tmp_path):
+    """
+    test execute delegate to allowed validator should succeed
+    test execute delegate to other validators should fail
+    """
+    # test execute delegate to allowed validator should succeed
+    spend_limit = 200
+    delegate_coins = 100
+    allowed_validator_address = cluster.validators()[0]["operator_address"]
+    another_validator_address = cluster.validators()[1]["operator_address"]
+    granter_address = cluster.address("community")
+    grantee_address = cluster.address("ecosystem")
+    validator_initial_deligation_amount = int(
+        query_delegation_amount(cluster, granter_address, allowed_validator_address)[
+            "amount"
+        ]
+    )
+
+    grant_authorization(
+        cluster,
+        grantee_address,
+        AUTHORIZATION_DELEGATE,
+        granter_address,
+        spend_limit="%s%s" % (spend_limit, BASECRO_DENOM),
+        allowed_validators=allowed_validator_address,
+    )
+    generated_delegate_txt = tmp_path / "generated_delegate.txt"
+    generated_delegate_msg1 = delegate_amount(
+        cluster,
+        allowed_validator_address,
+        "%s%s" % (delegate_coins, BASECRO_DENOM),
+        granter_address,
+        GENERATE_ONLY,
+    )
+    with open(generated_delegate_txt, "w") as opened_file:
+        json.dump(generated_delegate_msg1, opened_file)
+    exec_tx_by_grantee(cluster, generated_delegate_txt, grantee_address)
+
+    assert query_delegation_amount(
+        cluster, granter_address, allowed_validator_address
+    ) == {
+        "denom": BASECRO_DENOM,
+        "amount": str(validator_initial_deligation_amount + delegate_coins),
+    }
+
+    # test execute delegate to other validators not in allowed validators should fail
+    generated_delegate_msg2 = delegate_amount(
+        cluster,
+        another_validator_address,
+        "%s%s" % (delegate_coins, BASECRO_DENOM),
+        granter_address,
+        GENERATE_ONLY,
+    )
+    with open(generated_delegate_txt, "w") as opened_file:
+        json.dump(generated_delegate_msg2, opened_file)
+
+    with pytest.raises(Exception, match=r".*unauthorized.*"):
+        exec_tx_by_grantee(cluster, generated_delegate_txt, grantee_address)
+
+
+def test_unable_to_execute_delegate_to_deny_validator(cluster, tmp_path):
+    """
+    test execute delegate to deny validator should fail
+    """
+    spend_limit = 200
+    delegate_coins = 100
+    deny_validator_address = cluster.validators()[0]["operator_address"]
+    granter_address = cluster.address("community")
+    grantee_address = cluster.address("ecosystem")
+
+    grant_authorization(
+        cluster,
+        grantee_address,
+        AUTHORIZATION_DELEGATE,
+        granter_address,
+        spend_limit="%s%s" % (spend_limit, BASECRO_DENOM),
+        deny_validators=deny_validator_address,
```

Comment:
and is the logic that one can delegate to any validators except the ones on the deny list? not sure if the cluster in this test suite has multiple validators, but perhaps that could be good to check as well (one can't delegate to denied validators, but can delegate to others?)

Reply:
One cannot delegate to denied validators and any other validators that are not in allowed validators. I think this `deny-validator` flag is just stating some denied validators explicitly.